src/http_proxy/mod.rs

```diff
@@ -0,0 +1,226 @@
+use std::net::SocketAddr;
+use std::sync::Arc;
+use futures::sink::SinkExt;
+
+use http::{Request, Response};
+use openssl::pkey::{PKey, Private};
+use openssl::x509::X509;
+use tokio_util::codec::Framed;
+use tokio::net::{TcpListener, TcpStream};
+use tokio_native_tls::{TlsAcceptor, TlsStream};
+use tokio::stream::StreamExt;
+
+use crate::certificates::{load_key_from_file, native_identity, spoof_certificate, CA};
+use crate::codecs::http11::{HttpClient, HttpServer};
+use crate::SafeResult;
+
+pub mod mitm;
+pub use self::mitm::{MitmLayer, RequestCapture, ResponseCapture};
+
+use http::header::HeaderName;
+
+lazy_static! {
+    static ref CERT_AUTH: crate::certificates::CA =
+        CA::load_from_pem_files("ca/ca_certs/cert.pem", "ca/ca_certs/key.pem").unwrap();
+    static ref KEY: PKey<Private> = load_key_from_file("ca/ca_certs/key.pem").unwrap();
+}
+
+pub async fn start_mitm<T>(
+    port: u16,
+    mitm: Arc<T>,
+) -> Result<(), Box<dyn std::error::Error>>
+where
+    T: MitmLayer + std::marker::Sync + std::marker::Send,
+    T: 'static,
+{
+    let addr = format!("127.0.0.1:{}", port);
+    println!("mitm proxy listening on {}", addr);
+    let addr = addr.parse::<SocketAddr>()?;
+    let mut new_client_listener = TcpListener::bind(&addr).await?;
+
+    loop {
+        let (new_client_stream, _) = new_client_listener.accept().await?;
+        let mut transport = Framed::new(new_client_stream, HttpClient);
+        if let Some(proxy_opening_request) = transport.next().await {
+            match proxy_opening_request {
+                Ok(proxy_opening_request) => {
+                    if proxy_opening_request.method() == http::Method::CONNECT {
+                        tokio::spawn(tls_mitm_wrapper(
+                            transport,
+                            proxy_opening_request,
+                            mitm.clone(),
+                        ));
+                    }
+                }
+                Err(e) => {
+                    dbg!(e);
+                }
+            }
+        } else {
+            unimplemented!();
+        }
+    }
+}
+
+async fn tls_mitm_wrapper<T>(
+    client_stream: Framed<TcpStream, HttpClient>,
+    opening_request: Request<Vec<u8>>,
+    mitm: Arc<T>,
+) where
+    T: MitmLayer,
```

Comment:
Generally when you have a single bound for a T it's more readable to use `impl` syntax instead of generics.

`mitm: Arc<impl MitmLayer>`